### ISSUE TITLE
Pre trained Faster RCNN evaluation metrics all zero on pascal voc 2010 for Detection 

### ISSUE THREAD

**Issue by kevalmorabia97** [pytorch/vision#2095]
Hi Everyone,

I am training the pretrained Faster RCNN model on PASCAL VOC 2010 dataset for Object Detection by following this pytorch finetuning tutorial: [pytorch.org/tutorials/intermediate/torchvision_tutorial.html](https://pytorch.org/tutorials/intermediate/torchvision_tutorial.html)
I am only using RandomHorizontalFlip transformation while training. [Can anyone suggest how to do transforms on target box coordinates if I want to do Resizing/Cropping/etc?]

The model is created as follows:
```
backbone = torchvision.models.mobilenet_v2(pretrained=True).features
backbone.out_channels = 1280

anchor_generator = AnchorGenerator(sizes=((32, 64, 128, 256, 512),), aspect_ratios=((0.5, 1.0, 2.0),))
roi_pooler = torchvision.ops.MultiScaleRoIAlign(featmap_names=[0], output_size=7, sampling_ratio=2)

model = FasterRCNN(backbone, num_classes=21, rpn_anchor_generator=anchor_generator, box_roi_pool=roi_pooler)
```

I train the models using the code in the tutorial ([github.com/pytorch/vision/blob/master/references/detection/engine.py](https://github.com/pytorch/vision/blob/master/references/detection/engine.py)). Please note that batch_size for training and evaluation both is 1.
```
params = [p for p in model.parameters() if p.requires_grad]
optimizer = torch.optim.Adam(params, lr=0.005, weight_decay=0.0005)

for epoch in range(20):
    train_one_epoch(model, optimizer, train_loader, device, epoch, print_freq=1000)
    evaluate(model, val_loader, device=device)
```

The model trains without any error, but evaluation everytime gives all evalation metrics as 0. I trained for 10 epochs and for all epochs, the evaluation metrics are all 0 as shown below:
```
Epoch: [0]  [   0/4998]  eta: 2:16:43  lr: 0.005000  loss: 3.7245 (3.7245)  loss_classifier: 2.9938 (2.9938)  loss_box_reg: 0.0000 (0.0000)  loss_objectness: 0.7086 (0.7086)  loss_rpn_box_reg: 0.0221 (0.0221)  time: 1.6414  data: 0.1888  max mem: 0
Epoch: [0]  [1000/4998]  eta: 0:10:36  lr: 0.005000  loss: 0.6537 (1.6350)  loss_classifier: 0.3121 (0.9100)  loss_box_reg: 0.1036 (0.2444)  loss_objectness: 0.1734 (0.2508)  loss_rpn_box_reg: 0.0538 (0.2298)  time: 0.1417  data: 0.0194  max mem: 0
...
Epoch: [0]  [4997/4998]  eta: 0:00:00  lr: 0.005000  loss: 0.5429 (0.8217)  loss_classifier: 0.1480 (0.3901)  loss_box_reg: 0.0584 (0.1153)  loss_objectness: 0.1542 (0.1866)  loss_rpn_box_reg: 0.0500 (0.1296)  time: 0.1659  data: 0.0356  max mem: 0
Epoch: [0] Total time: 0:12:43 (0.1527 s / it)
creating index...
index created!
Test:  [   0/5105]  eta: 5:00:57  model_time: 0.0213 (0.0213)  evaluator_time: 0.0009 (0.0009)  time: 3.5373  data: 0.0073  max mem: 0
Test:  [1000/5105]  eta: 0:02:09  model_time: 0.0188 (0.0189)  evaluator_time: 0.0008 (0.0008)  time: 0.0282  data: 0.0073  max mem: 0
...
Test:  [5104/5105]  eta: 0:00:00  model_time: 0.0191 (0.0192)  evaluator_time: 0.0008 (0.0011)  time: 0.0311  data: 0.0100  max mem: 0
Test: Total time: 0:03:29 (0.0411 s / it)
Averaged stats: model_time: 0.0191 (0.0192)  evaluator_time: 0.0008 (0.0011)
Accumulating evaluation results...
DONE (t=0.85s).
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
```
Can anyone please help me resolve the issue?

Thanks for your time!

**Comment by fmassa** [comment]
Hi,

This can be due to a number of factors.
You are training the model from scratch, which might require a lot of training iterations before it starts giving meaningful results, specially because Pascal is small.

I think the key issue here though is that there might be an issue with your dataset implementation, which is not returning things in the format the the evaluation code expects.

I would visualize the output of the model after the 10 epochs that you currently have and see if they start to look reasonable or not yet.

Without further information, it's difficult to say much more about it.

**Comment by kevalmorabia97** [comment]
Hi @fmassa, I just updated my answer above. I was able to resolve to issue of all metrics to be 0 by changing from mobilenet_v2 backbone to the default backbone, and optimizer from Adam to SGD. I still am getting much worse results then expected (47% vs 70% mAP). Could you please help me out?

**Comment by fmassa** [comment]
I would say to increase the number of training epochs and the learning rate as a start.
Also, increasing the training set size (by using trainval instead of train) will definitely improve performances.

**Comment by WZMIAOMIAO** [comment]
hi, my personal opinion is that your result is good. The mAP in faster R-CNN is  69.9 when iou = 0.5, not IoU=0.50:0.95. in your result, when iou = 0.5, mAP is 76.8.

**Comment by kevalmorabia97** [comment]
Thanks @WZMIAOMIAO for pointing out my mistake!
I was able to get even better results by reducing weight decay to 1e-6 and adding learning rate decay.

After training on ```VOC 2010 train``` for 10 epochs, I got ```mAP[@IoU=0.5] = 79.4``` on ```VOC 2010 val```.
After training on ```VOC 2007 trainval``` for 10 epochs, I got ```mAP[@IoU=0.5] = 86.0``` on ```VOC 2007 test```.
